components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodTerminalService.kt

```diff
@@ -25,41 +28,34 @@ import java.util.concurrent.CompletableFuture
 import java.util.concurrent.TimeUnit
 
 @Suppress("UnstableApiUsage")
-class GitpodTerminalService(session: ClientProjectSession) : Disposable {
+class GitpodTerminalService(private val session: ClientProjectSession) {
     private companion object {
-        /** Indicates if this service is already running, because we shouldn't run it more than once. */
-        var isRunning = false
+        var hasCreatedTerminals = false
+        val forwardedPortsList: MutableSet<Int> = mutableSetOf()
     }
 
-    private val lifetime = Lifetime.Eternal.createNested()
     private val terminalView = TerminalView.getInstance(session.project)
     private val backendTerminalManager = BackendTerminalManager.getInstance(session.project)
+    private val portForwardingManager = PortForwardingManager.getInstance(session.project)
     private val terminalServiceFutureStub = TerminalServiceGrpc.newFutureStub(GitpodManager.supervisorChannel)
     private val statusServiceStub = StatusServiceGrpc.newStub(GitpodManager.supervisorChannel)
 
-    override fun dispose() {
-        lifetime.terminate()
-    }
-
-    init {
-        run()
-    }
-
-    private fun run() {
-        if (application.isHeadlessEnvironment || isRunning) return
+    init { start() }
 
-        isRunning = true
+    private fun start() {
+        if (application.isHeadlessEnvironment || hasCreatedTerminals) return
 
         val task = application.executeOnPooledThread {
```

Comment:
I wonder do we really need to cancel it? maybe removing cancellation is the easiest? then disconnections cannot cause incomplete terminals and so on?

As I read this code it still can race if you reconnect before `hasCreatedTerminals` is set?